core/primitives-core/src/types.rs

```diff
@@ -45,3 +45,70 @@ pub type ReceiptIndex = usize;
 pub type PromiseId = Vec<ReceiptIndex>;
 
 pub type ProtocolVersion = u32;
+
+/// The shard identifier. It may be a arbitrary number - it does not need to be
+/// a number in the range 0..NUM_SHARDS. The shard ids do not need to be
+/// sequential or contiguous.
+///
+/// The shard id is wrapped in a newtype to prevent the old pattern of using
+/// indices in range 0..NUM_SHARDS and casting to ShardId. Once the transition
+/// if fully complete it potentially may be simplified to a regular type alias.
+///
+/// TODO get rid of serde
```

Comment:
What prevents removing serde now?

Reply:
Some tests use serde as a snapshotting library to make sure old layouts are not
changed accidentally.